modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Logic/GetErc20Balance.swift

```diff
@@ -4,24 +4,41 @@ import Foundation
 import BigInt
 import PromiseKit
 import AlphaWalletWeb3
+import AlphaWalletCore
 
-public class GetErc20Balance {
+class GetErc20Balance {
     private let server: RPCServer
-    private let queue: DispatchQueue?
+    private var inFlightPromises: [AlphaWallet.Address: Promise<BigInt>] = [:]
+    private let queue = DispatchQueue(label: "org.alphawallet.swift.getErc20Balance")
 
-    public init(forServer server: RPCServer, queue: DispatchQueue? = nil) {
+    init(forServer server: RPCServer) {
         self.server = server
-        self.queue = queue
     }
 
-    public func getBalance(for address: AlphaWallet.Address, contract: AlphaWallet.Address) -> Promise<BigInt> {
-        let functionName = "balanceOf"
-        return callSmartContract(withServer: server, contract: contract, functionName: functionName, abiString: Web3.Utils.erc20ABI, parameters: [address.eip55String] as [AnyObject], queue: queue)
-            .map(on: queue, { balanceResult in
-                guard let balanceOfUnknownType = balanceResult["0"], let balance = BigInt(String(describing: balanceOfUnknownType)) else {
-                    throw CastError(actualValue: balanceResult["0"], expectedType: BigInt.self)
-                }
-                return balance
-            })
+    func getBalance(for address: AlphaWallet.Address, contract: AlphaWallet.Address) -> Promise<BigInt> {
+        firstly {
+            .value(contract)
+        }.then(on: queue, { [weak self, queue, server] contract -> Promise<BigInt> in
+            if let promise = self?.inFlightPromises[contract] {
```

Comment:
Can't cache with just `contract`, I think it needs at least `contract+address/wallet` as the key. Similar for at least some of the rest?

Reply:
oh, yes right, updating